spark2/src/main/java/org/apache/iceberg/spark/source/Writer.java

```diff
@@ -250,33 +253,42 @@ public String toString() {
       if (spec.fields().isEmpty()) {
         return new Unpartitioned24Writer(spec, format, appenderFactory, fileFactory, io.value(), targetFileSize);
       } else {
-        return new Partitioned24Writer(
-            spec, format, appenderFactory, fileFactory, io.value(), targetFileSize, writeSchema);
+        return new Partitioned24Writer(spec, format, appenderFactory, fileFactory, io.value(),
+            targetFileSize, writeSchema, dsSchema);
       }
     }
   }
 
-  private static class Unpartitioned24Writer extends UnpartitionedWriter implements DataWriter<InternalRow> {
+  private static class Unpartitioned24Writer extends UnpartitionedWriter<InternalRow>
+      implements DataWriter<InternalRow> {
     Unpartitioned24Writer(PartitionSpec spec, FileFormat format, SparkAppenderFactory appenderFactory,
                           OutputFileFactory fileFactory, FileIO fileIo, long targetFileSize) {
       super(spec, format, appenderFactory, fileFactory, fileIo, targetFileSize);
     }
 
     @Override
     public WriterCommitMessage commit() throws IOException {
-      return new TaskCommit(complete());
+      this.close();
+
+      List<DataFile> dataFiles = complete();
+      return new TaskCommit(new TaskResult(dataFiles));
```

Comment:
If `complete` doesn't produce `TaskResult`, then I'm not sure that we need it at all anymore. Could we just construct `TaskCommit` directly?

Reply:
OK